test/unit/org/apache/cassandra/cql3/validation/entities/TupleTypeTest.java

```diff
@@ -233,5 +256,149 @@ public void testReversedTypeTuple() throws Throwable
+            for (int i = 0; !done && i < numRows; i++)
+            {
+                for (int retry = 0; retry < 10 && !distinctRows.add(valueGen.generate(rnd)); retry++)
+                {
+                    if (retry == 9)
+                    {
+                        System.err.println("Not enough uniq row for type " + c.type + ", returning with " + i + "rows");
+                        done = true;
+                        break;
+                    }
+                }
+            }
+            c.uniqueRows = new ArrayList<>(distinctRows);
+            return c;
+        };
+        gen = gen.describedAs(c -> c.type.asCQL3Type().toString());
+        qt().withExamples(10).withShrinkCycles(0).forAll(gen).checkAssert(orFail(testcase -> {
+            TupleType tupleType = testcase.type;
+            createTable("CREATE TABLE %s (id " + toCqlType(tupleType) + ", value int, PRIMARY KEY(id))");
+            SortedMap<ByteBuffer, Integer> map = new TreeMap<>(Comparator.comparing(currentTableMetadata().partitioner::decorateKey));
+            int count = 0;
+            for (ByteBuffer value : testcase.uniqueRows)
+            {
+                map.put(value, count);
+                ByteBuffer[] tupleBuffers = tupleType.split(value);
+
+                execute("INSERT INTO %s (id, value) VALUES (?, ?)", tuple(tupleBuffers), count);
+
+                assertRows(execute("SELECT * FROM %s WHERE id = ?", tuple(tupleBuffers)),
+                           row(tuple(tupleBuffers), count));
+                count++;
+            }
+            assertRows(execute("SELECT * FROM %s LIMIT 100"),
+                       map.entrySet().stream().map(e -> row(e.getKey(), e.getValue())).toArray(Object[][]::new));
+        }));
+    }
+
+    @Test
+    public void tupleCkReadWriteAsc()
+    {
+        tupleCkReadWrite(Order.ASC);
+    }
+
+    @Test
+    public void tupleCkReadWriteDesc()
+    {
+        tupleCkReadWrite(Order.Desc);
+    }
+
+    private void tupleCkReadWrite(Order order)
+    {
+        int numRows = 10;
+        class Case {
+            TupleType type;
+            List<ByteBuffer> uniqueRows;
+        }
+        Gen<TupleType> baseGen = getTupleType(getPrimitiveType(), SourceDSL.integers().between(1, 10));
+        Set<ByteBuffer> distinctRows = new HashSet<>(); // reuse the memory
+        Gen<Case> gen = rnd -> {
+            Case c = new Case();
+            c.type = baseGen.generate(rnd);
+            TypeSupport<ByteBuffer> support = getTypeSupport(c.type);
+            Gen<ByteBuffer> valueGen = Generators.filter(support.valueGen, b -> b.remaining() <= Short.MAX_VALUE);
+
+            distinctRows.clear();
+            for (int i = 0; i < numRows; i++)
+            {
+                while (!distinctRows.add(valueGen.generate(rnd))) { }
+            }
+            c.uniqueRows = new ArrayList<>(distinctRows);
+            return c;
+        };
+        gen = gen.describedAs(c -> c.type.asCQL3Type().toString());
```

Comment:
On failure you wouldn't get to see the `uniqueRows`.

Reply:
true, was mostly looking at the schema when the test was failing, so didn't want the extra noise.